dot/rpc/modules/grandpa.go

```diff
@@ -71,3 +97,101 @@ func (gm *GrandpaModule) ProveFinality(r *http.Request, req *ProveFinalityReques
 
 	return nil
 }
+
+// RoundState returns the state of the current best round state as well as the ongoing background rounds.
+func (gm *GrandpaModule) RoundState(r *http.Request, req *EmptyRequest, res *RoundStateResponse) error {
+	voters := gm.blockFinalityAPI.GetVoters()
+	votersPkBytes := make([]ed25519.PublicKeyBytes, len(voters))
+	for i, v := range voters {
+		votersPkBytes[i] = v.PublicKeyBytes()
+	}
+
+	prevotes := gm.blockFinalityAPI.PreVotes()
+	precommits := gm.blockFinalityAPI.PreCommits()
+
+	missingPrevotes, err := toAddress(difference(votersPkBytes, prevotes))
+	if err != nil {
+		return err
+	}
+
+	missingPrecommits, err := toAddress(difference(votersPkBytes, precommits))
+	if err != nil {
+		return err
+	}
+
+	totalWeight := calcTotalWeight(voters)
+
+	roundstate := RoundStateResponse{
+		SetID: uint32(gm.blockFinalityAPI.GetSetID()),
+		Best: RoundState{
+			Round:           uint32(gm.blockFinalityAPI.GetRound()),
+			ThresholdWeight: calcThresholdWeight(totalWeight),
+			TotalWeight:     totalWeight,
+			Prevotes: Votes{
+				CurrentWeight: calcWeight(voters, prevotes),
+				Missing:       missingPrevotes,
+			},
+			Precommits: Votes{
+				CurrentWeight: calcWeight(voters, precommits),
+				Missing:       missingPrecommits,
+			},
+		},
+		Background: []RoundState{},
+	}
+
+	*res = roundstate
+	return nil
+}
+
+func calcWeight(voters grandpa.Voters, pre map[ed25519.PublicKeyBytes]*grandpa.Vote) uint32 {
+	var weight uint32
+	for pk := range pre {
+		for _, gpv := range voters {
+			if gpv.PublicKeyBytes() == pk {
+				weight += uint32(gpv.ID)
+			}
+		}
+	}
+	return weight
+}
```

Comment:
if current weight is supposed to be the weight of the current votes for a subround, then it can just be `len(votes)` for a specific subround

Reply:
Yes, I think it should be. Can I use `len(precommits)` and `len(prevotes)` to the fields `Precommits.CurrentWeight` and `Prevotes.CurrentWeight` respectively, rather than use the `calcWeight` function?

Reply:
yes, that is fine!

Reply:
actually, since you added the equivocations, you'll need to check how many distinct authorities actually voted, since equivocations contains multiple conflicting votes from the same authority

Reply:
Ok, so basically I need to check if the voter is inside the equivocation list, if true then I will consider the voter as missing?

Reply:
if the voter is in the equivocation list, they have already voted so they would not be missing. only voters that are neither in the votes or equivocations lists would be missing

Reply:
Got it, I will update the logic! I thought that the equivocation list was a set of indecision voters.